### ISSUE TITLE
[MNG-5937] Maven-Wrapper for unified project environments (analog the Gradle wrapper)

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7747]
**[Robert Weiser](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=death-knight)** opened **[MNG-5937](https://issues.apache.org/jira/browse/MNG-5937?redirect=false)** and commented

It would be nice to provide a wrapper for maven which behaves similarily to the wrapper Gradle has. This would allow projects to share Maven in a certain version along with some project-specific Maven options like for example the location of the user or the global settings files or the maximum java heap size to use.

There already are projects on github (e.g.: https://github.com/takari/maven-wrapper) providing said feature, however it would be nice to have it developed and maintained by the very people providing us with the offical Maven release.

Cheers, Robert

---

**Issue Links:**
- [MWRAPPER-17](https://issues.apache.org/jira/browse/MWRAPPER-17) IP-CLEARANCE for Maven Wrapper
 (_**"is blocked by"**_)
- [MNG-7136](https://issues.apache.org/jira/browse/MNG-7136) Update Maven modules structure schema to add new components
 (_**"incorporates"**_)
- [MWRAPPER-14](https://issues.apache.org/jira/browse/MWRAPPER-14) Make maven wrapper functional
 (_**"is superceded by"**_)
- [MNG-6914](https://issues.apache.org/jira/browse/MNG-6914) Align mvn and mvnw scripts
 (_**"is depended upon by"**_)

9 votes, 12 watchers

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

It is developed by the people who do the Maven releases. That would be me.

**Comment by jira-importer** [comment]
**[Robert Weiser](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=death-knight)** commented

Shame on me - I should have taken the time to check the official members list.

Are there any plans to provide the maven-wrapper as an additional download alternative?

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

At some point, it will most likely be integrated. I have a very long blog entry about why it's not. Along with Polyglot Maven and many of the other Takari extensions. Primary reasons are that I prefer the Github infrastructure, and in many cases users have contributed PRs and I can integrate them and release them an hour later because it's critical for that user. I simply can't do that at Apache and I believe it's a fundamental detractor.

**Comment by jira-importer** [comment]
**[Fabrizio Cucci](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=fabrizio.cucci)** commented

It's a shame that such a great feature is not really "official" like it is for Gradle.

[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl) could you please point out the blog entry you mentioned before?

I'm really curious to know why this issue is still open in 2017...

**Comment by jira-importer** [comment]
**[Jiří Pejchal](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jiri_pejchal)** commented

When the maven wrapper is an external project IDEs are reluctant to add support for it. There is no support for maven wrapper in intellij idea https://youtrack.jetbrains.com/issue/IDEA-146205 but gradle wrapper is supported.

**Comment by jira-importer** [comment]
**[Adrian Cole](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=adriancole)** commented

I would consider the maven wrapper a backdoor required feature. I like it, but it is troubling that the team is not able to donate this to the ASF, and even are going so far as to suggest every ASF project needs to advertise it in their NOTICE file. We don't have this behavior even from gradle.. maybe they would like to have projects add GradleWare to the NOTICE file, too? Maybe they will one day!

https://github.com/takari/takari-maven-plugin/issues/18#issuecomment-464867539

Anyway, maven should be fully functional, and I'd argue that at this point wrapper is a part of it being functional. I think the community should be at least permitted to write their own version of the wrapper if the takari folks are unwilling.

**Comment by jira-importer** [comment]
**[Adrian Cole](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=adriancole)** commented

Also, I would suggest that 3+ years later, maybe the wrapper is sufficiently mature to afford ASF release process. It might not have been the case in 2015.. some things change over time.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Completed at [cdcd351928a6dc628cbb74a792fe6fb6589e3130](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=cdcd351928a6dc628cbb74a792fe6fb6589e3130)
with integration tests at [077b9d3c9d9efcd9192082516d358b943ea3edcf](https://gitbox.apache.org/repos/asf?p=maven-integration-testing.git;a=commit;h=077b9d3c9d9efcd9192082516d358b943ea3edcf)

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

https://github.com/search?q=repo%3Aapache%2Fmaven+MNG-5937&type=commits

superceded by MWRAPPER-14 https://github.com/search?q=repo%3Aapache%2Fmaven+MWRAPPER-14&type=commits
